### PR TITLE
memtable flush reserve process mem and improve logs

### PR DESCRIPTION
**Pull request by jacktengg**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by Thearas** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 1]
```diff
@@ -278,34 +278,44 @@ inline void ThreadMemTrackerMgr::flush_untracked_mem() {
     _stop_consume = false;
 }
 
-inline doris::Status ThreadMemTrackerMgr::try_reserve(int64_t size) {
+inline doris::Status ThreadMemTrackerMgr::try_reserve(int64_t size,
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:281]: 把这个API 的单测补充一下

File: be/src/runtime/thread_context.h [hunk 2]
```diff
@@ -249,13 +249,22 @@ class ThreadContext {
         thread_mem_tracker_mgr->consume(size, skip_large_memory_check);
     }
 
+    doris::Status reserve_memory(const int64_t size) const {
```
Comment (yiguolei) [inline be/src/runtime/thread_context.h:252]: xxx 和 try_xxx 区分不清楚是否要检查query和wg的limit。
我们换个名字，增加一个新的函数，try_reserve_process_memory

File: be/src/olap/memtable_flush_executor.cpp [hunk 3]
```diff
@@ -140,6 +141,36 @@ Status FlushToken::wait() {
...
+            break;
+        }
+        if (_is_shutdown()) {
+            st = Status::Cancelled("flush memtable already cancelled");
+            break;
+        }
+        // Make sure at least one memtable is flushing even reserve memory failed.
+        if (memtable_flush_executor->check_and_inc_has_any_flushing_task()) {
+            // If there are already any flushing task, Wait for some time and retry.
+            LOG(INFO) << fmt::format(
```
Comment (yiguolei) [inline be/src/olap/memtable_flush_executor.cpp:163]: 这个每1min 打印一次吧，否则内存满的时候刷屏了

File: be/src/olap/memtable_flush_executor.cpp [hunk 4]
```diff
@@ -140,6 +141,36 @@ Status FlushToken::wait() {
...
+    Status st;
+    do {
+        // only try to reserve process memory
+        st = thread_context->reserve_memory(size);
+        if (st.ok()) {
+            memtable_flush_executor->inc_flushing_task();
+            break;
+        }
+        if (_is_shutdown()) {
+            st = Status::Cancelled("flush memtable already cancelled");
```
Comment (yiguolei) [inline be/src/olap/memtable_flush_executor.cpp:157]: 如果在等待期间，要flush的这个memtable 关联的导入任务呗cancel了，怎么处理？

File: be/src/olap/memtable_flush_executor.cpp [hunk 5]
```diff
@@ -150,10 +181,18 @@ Status FlushToken::_do_flush_memtable(MemTable* memtable, int32_t segment_id, in
...
     signal::set_signal_task_id(_rowset_writer->load_id());
     signal::tablet_id = memtable->tablet_id();
+
+    DEFER_RELEASE_RESERVED();
+
     {
         SCOPED_SWITCH_THREAD_MEM_TRACKER_LIMITER(
                 memtable->query_thread_context().query_mem_tracker->write_tracker());
         SCOPED_CONSUME_MEM_TRACKER(memtable->mem_tracker());
+        auto reserve_size = memtable->get_flush_reserve_memory_size();
```
Comment (yiguolei) [inline be/src/olap/memtable_flush_executor.cpp:191]: 由于在188 行有一个scope switch了，所以185 行那个defer release 就不对了。
因为192 行作用在188 那个memtracker 上了，但是释放的时候就到了181 行。

File: be/src/vec/exec/scan/scanner_scheduler.cpp [hunk 6]
```diff
@@ -254,6 +287,14 @@ void ScannerScheduler::_scanner_scan(std::shared_ptr<ScannerContext> ctx,
             }
 
             if (!eos && !scanner->is_open()) {
+                if (state->enable_reserve_memory()) {
+                    size_t block_avg_bytes = scanner->get_block_avg_bytes();
+                    auto st = thread_context()->try_reserve_memory(block_avg_bytes);
```
Comment (yiguolei) [inline be/src/vec/exec/scan/scanner_scheduler.cpp:292]: 这里reserve之后没有释放吗？

File: be/src/vec/exec/scan/scanner_scheduler.cpp [hunk 7]
```diff
@@ -254,6 +287,14 @@ void ScannerScheduler::_scanner_scan(std::shared_ptr<ScannerContext> ctx,
...
+                    if (!st.ok()) {
+                        handle_reserve_memory_failure(state, ctx, st, block_avg_bytes);
```
Comment (yiguolei) [inline be/src/vec/exec/scan/scanner_scheduler.cpp:294]: 这里不能加这个，加了这个有可能会导致整个查询卡死。
我们现在scanner的调度，需要保证至少scanner 扫描出一个block

File: be/src/vec/exec/scan/scanner_scheduler.cpp [hunk 8]
```diff
@@ -268,16 +301,17 @@ void ScannerScheduler::_scanner_scan(std::shared_ptr<ScannerContext> ctx,
             }
 
             size_t raw_bytes_threshold = config::doris_scanner_row_bytes;
-            if (ctx->low_memory_mode() &&
-                raw_bytes_threshold > ctx->low_memory_mode_scan_bytes_per_scanner()) {
-                raw_bytes_threshold = ctx->low_memory_mode_scan_bytes_per_scanner();
+            if (ctx->low_memory_mode()) {
```
Comment (yiguolei) [inline be/src/vec/exec/scan/scanner_scheduler.cpp:304]: 这里有clear 了，280 行，还有意义吗？

Comment (jacktengg) [reply be/src/vec/exec/scan/scanner_scheduler.cpp:304]: 这中间的scanner->open 可能也会申请很多内存

**Review by yiguolei — APPROVED** [review]

**Pull request merged**